### ISSUE TITLE
Guest-Specific Operations Error Configuring Shared Folders

### ISSUE THREAD

**Issue by uchagani** [mitchellh/vagrant#5226]
I'm seeing the ```Guest-Specific Operations Error``` every now and then and only when creating a shared folder.  A few reboots usually fix the issue but it's very inconvenient to reboot a large server a couple times a day.

Looking through the logs it looks like Vagrant is unable to get a proper IP Address from ```vmrun``` when creating folder shares.  I've only seen this during folder share creation.  All the stuff before that gets the correct IP Address.

It seems after getting a bad IP Address, Vagrant tries to run the ```hostname``` command but nothing is returned because it's connecting over a bad IP.

Respective Area of the debug log:

```
 INFO winrm: Checking whether WinRM is ready...
DEBUG provider: Searching for cap: winrm_info
DEBUG provider: Checking in: vmware_workstation
 INFO subprocess: Starting process: ["C:\\Program Files (x86)\\VMware\\VMware Workstation/vmrun.exe", "list"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: Total running VMs: 1
D:\pipeline\vms\test1\.vagrant\machines\default\vmware_workstation\2750adf3-e798-4560-82da-d4383cc07c62\IMSBaseBox.vmx
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 31999
DEBUG subprocess: Exit status: 0
DEBUG vmware: VM state requested. Current state: running
 INFO vmware_driver: Reading an accessible IP for machine...
 INFO vmware_driver: Trying vmrun getGuestIPAddress...
 INFO subprocess: Starting process: ["C:\\Program Files (x86)\\VMware\\VMware Workstation/vmrun.exe", "getGuestIPAddress", "D:/pipeline/vms/test1/.vagrant/machines/default/v
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: 169.254.107.135
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
 INFO vmware_driver: vmrun getGuestIPAddress success: 169.254.107.135
DEBUG winrmshell: initializing WinRMShell
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO winrmshell: Attempting to connect to WinRM...
 INFO winrmshell:   - Host: 169.254.107.135
 INFO winrmshell:   - Port: 5985
 INFO winrmshell:   - Username: vagrant
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO retryable: Retryable exception raised: #<Errno::ENETUNREACH: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
DEBUG winrmshell: powershell executing:
hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
 INFO winrm: WinRM not up: #<VagrantPlugins::CommunicatorWinRM::Errors::ExecutionError: An error occurred executing a remote WinRM command.

Shell: powershell
Command: hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
Message: A socket operation was attempted to an unreachable network. - connect(2) (http://169.254.107.135:5985)>
ERROR warden: Error occurred: Guest-specific operations were attempted on a machine that is not
ready for guest communication. This should not happen and a bug
should be reported.
```

**Comment by uchagani** [comment]
@mitchellh @sethvargo I've come up concrete reproducible steps that are a combination of base box network configurations and vagrant.  Are either of you available to chat on IRC or such to further explain this.  I think I've identified a way to fix this issue plus the whole ```Configuring secondary network adapters through VMWare on Windows is not yet supported.```

**Comment by mitchellh** [comment]
@uchagani What is your fix? Care to expand?

**Comment by uchagani** [comment]
Regarding ```Guest-Specific Operations``` Error:

This error is thrown when a Windows Guest is configured with a public network and gets a 169.254.xxx address from the DHCP server.  Since Vagrant only looks for a IP Address in the correct format, when the guest receives a 169.254 address, vagrant attempts to perform guest operations on the Guest, causing this error to get thrown.  This is not a vagrant or VMWare problem; it is a Windows problem.  The way to solve it is to keep waiting until you receive an IP Address that is not a 169.254 address.  I was able to reproduce the ```guest-specific operations error``` consistently by creating a basebox that already had a public NIC configured.


Regarding ```Configuring secondary network adapters through VMWare on Windows is not yet supported``` error:

Vagrant says that it can't configure secondary networks but configuring secondary networks is pretty simple by modifying the ```vmx``` file.  I've been able use vagrant to modify the ```vmx``` file with success.

**Comment by uchagani** [comment]
Regarding the ```Configuring secondary network adapters through VMWare on Windows is not yet supported``` see #5000 for an example of how to do this.

**Comment by StefanScherer** [comment]
:+1: I would love to see secondary network adapters in VMware for Windows guests. Lots of my multi machine Vagrantfile work only with VirtualBox at the moment.

How about adding a private network instead of a bridged one? Is it the same way in the vmx file?

**Comment by uchagani** [comment]
@StefanScherer , I believe so.  All you need to do is look at the VMX file to figure out what the values need to be, then use vagrant to modify the VMX file.

**Comment by mitchellh** [comment]
Given this is a Windows issue @uchagani, I'm unfortunately going to have to close this until we can come up with a PR for a fix for this since I don't really have the expertise to do this and this issue will rot otherwise. Are you or @StefanScherer knowledgable for how to do this? Let me know if I can help!

**Comment by StefanScherer** [comment]
@mitchellh please have a look at my last comment in https://github.com/mitchellh/vagrant/issues/5000#issuecomment-158742386 to set the static IP address of a second network adapter in a Windows VM. I hope that will help you.

**Comment by f3rrix** [comment]
Not sure whether this helps in the meanwhile:
I noticed in the code there are 2 methods to get the IP.  The vmrun getGuestIPAddress, which is what returns the wrong answer, and a parsing of the vmx file which returns the correct (NAT) address.

So I hacked subprocess.rb to prohibit the getGuestIPAddress command from running, thus forcing it to use the method which always works.
[subprocess.rb.1.7.1.patch.txt](https://github.com/mitchellh/vagrant/files/170800/subprocess.rb.1.7.1.patch.txt)

(Also, unrelatedly, told it to shut up if shared folders are already enabled instead of it being an error)
